spec/lib/linters/crunch_api/report_retriever_spec.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+require 'rest-client'
+
+describe Linters::CrunchApi::ReportRetriever do
+  let(:rest_client) { class_spy(RestClient, get: 'test_report') }
+  let(:crunch_base_url) { 'https://api.crunch.com' }
+  let(:api_id) { '123' }
+  let(:api_key) { 'api_key' }
+
+  describe '#retrieve_report_for_api' do
+    let(:url) { "#{crunch_base_url}/api/v1/apis/#{api_id}/assessmentreport" }
+    let(:response) { double('response') }
+
+    before do
+      stub_const('Linters::CrunchApi::CRUNCH_BASE_URL', crunch_base_url)
+      stub_const('Linters::CrunchApi::CRUNCH_API_KEY', api_key)
+    end
+
+    context 'when the crunch API returns a 200' do
+      it 'returns the response from the API' do
+        report_retriever = described_class.new(rest_client:, base_url: crunch_base_url)
+
+        expect(report_retriever.retrieve_report_for_api(api_id)).to eq('test_report')
+        expect(rest_client).to have_received(:get).with(url, { 'X-API-KEY': api_key })
```

Comment:
Webmock is not required in a proper unit test like this. Adding webmock means we are testing more than what the report retriever does and is bad practise for unit testing as you are turning it into an integration test. 

We have a request spec that uses VCR which uses webmock behind the scenes to test the full flow of the code with its dependencies.

Reply:
I'm a little confused by your argument. Surely fundamentally what this object does is send a request out to a remote API and process the response. Therefore that is precisely what the unit test should test.

I would agree that if another object or process created the response which was passed to this object to process, then it would be inappropriate to mock the other objects/process' callout to the API. Rather a mock response would be passed to this object to test. But I don't believe that is the case here.

What adding spies to `RestClient` does is tie this object to using that service. It adds a dependency to use `RestClient`. If webmock is used that dependency is broken and the object could use anything to make the call: Faraday, HTTParty, Ruby's HTTP library. 

However, as I said previously I don't agree with you but will be happy to approve this PR with the spying and mocking strategy in place.